### PR TITLE
gluon-status-page, gluon-mesh-batman-adv: make statuspage mesh-protoc…

### PR DESCRIPTION
**Pull request by christf**
…ol-agnostic

this PR will make the staus page backend independent from batctl.

this replaces the backend-part from #1053

### CODE REVIEW

File: package/gluon-mesh-batman-adv/src/Makefile [hunk 1]
```diff
@@ -33,3 +33,4 @@ LDLIBS += $(LIBBATADV_LDLIBS)
 
 respondd.so: respondd.c
 	$(CC) $(CPPFLAGS) $(CFLAGS) $(LDFLAGS) -shared -fPIC -D_GNU_SOURCE -o $@ $^ $(LDLIBS) -lgluonutil -liwinfo -luci
+
```
Comment (NeoRaider) [inline package/gluon-mesh-batman-adv/src/Makefile]: It seems this line was added accidentally?

File: package/gluon-status-page-mesh-batman-adv/Makefile [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+PKG_NAME:=gluon-status-page-mesh-batman-adv
+PKG_VERSION:=1
+
+include ../gluon.mk
+
+define Package/gluon-status-page-mesh-batman-adv
+  TITLE:=Batman-data provider for gluon-status-page
+  SECTION:=gluon
+  CATEGORY:=Gluon
```
Comment (NeoRaider) [inline package/gluon-status-page-mesh-batman-adv/Makefile]: The above two lines can be dropped since 994c94918a73a673f25a65fd6795fd89e19decec

File: package/gluon-status-page/files/lib/gluon/status-page/www/cgi-bin/dyn/neighbours-nodeinfo [hunk 3]
```diff
@@ -6,6 +6,6 @@ badrequest() {
         exit 1
 }
 
-( batctl if | cut -d: -f1 | grep -qxF "$QUERY_STRING" ) 2>/dev/null || badrequest
+( ubus call network.interface dump | jsonfilter -e "@.interface[@.proto='gluon_mesh' && @.up=true].device"| grep -qxF "$QUERY_STRING" ) 2>/dev/null || badrequest
```
Comment (NeoRaider) [inline package/gluon-status-page/files/lib/gluon/status-page/www/cgi-bin/dyn/neighbours-nodeinfo]: Hmm, I think the complexity of this command is high enough that we should encapsulate the whole thing up to the `||` in a small script provided by gluon-core. Something like `gluon-is-meshif`. I guess the `grep` could also be replaced by an extension of the jsonfilter expression.

File: package/gluon-status-page-mesh-batman-adv/src/Makefile [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  endif
+  LIBBATADV_CFLAGS += $(shell $(PKG_CONFIG) --cflags $(LIBBATADV_NAME))
+  LIBBATADV_LDLIBS += $(shell $(PKG_CONFIG) --libs $(LIBBATADV_NAME))
+endif
+CFLAGS += $(LIBBATADV_CFLAGS)
+LDLIBS += $(LIBBATADV_LDLIBS)
+
+neighbours-batadv: neighbours-batadv.c
+	$(CC) $(CFLAGS) $(CFLAGS_JSONC) $(LDFLAGS) $(LDFLAGS_JSONC) -o $@ $^ $(LDLIBS)
+
```
Comment (NeoRaider) [inline package/gluon-status-page-mesh-batman-adv/src/Makefile]: Ugh, I hate the amount of boilerplate code in this file... I'll clean this up very soon, in all Makefiles that use libbatadv.

File: package/gluon-status-page/src/stations.c [hunk 5]
```diff
@@ -27,6 +27,8 @@ static struct json_object *get_stations(const struct iwinfo_ops *iw, const char
     json_object_object_add(station, "signal", json_object_new_int(entry->signal));
     json_object_object_add(station, "noise", json_object_new_int(entry->noise));
     json_object_object_add(station, "inactive", json_object_new_int(entry->inactive));
+    json_object_object_add(station, "rx_rate", json_object_new_int(entry->rx_rate.rate));
+    json_object_object_add(station, "tx_rate", json_object_new_int(entry->tx_rate.rate));
```
Comment (NeoRaider) [inline package/gluon-status-page/src/stations.c]: This change looks unrelated, and we should not add these values unless the frontend also uses them.

File: package/gluon-status-page-mesh-batman-adv/Makefile [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  PROVIDES:=gluon-status-page-data-provider
+  DEPENDS:=+gluon-status-page +gluon-mesh-batman-adv +libnl-tiny +libbatadv +libjson-c
```
Comment (NeoRaider) [inline package/gluon-status-page-mesh-batman-adv/Makefile]: *gluon-mesh-batman-adv* is a virtual package. Remove the + sign prefixing the dependency - never use selecting dependencies with virtual packages!

File: package/gluon-status-page-mesh-batman-adv/Makefile [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+PKG_NAME:=gluon-status-page-mesh-batman-adv
+PKG_VERSION:=1
+
+include ../gluon.mk
+
+define Package/gluon-status-page-mesh-batman-adv
+  TITLE:=Batman-data provider for gluon-status-page
+  SECTION:=gluon
+  CATEGORY:=Gluon
+  PROVIDES:=gluon-status-page-data-provider
```
Comment (NeoRaider) [inline package/gluon-status-page-mesh-batman-adv/Makefile]: I think this PROVIDES is somewhat useless; as long as *gluon-status-page-mesh-batman-adv* depends on *gluon-status-page*, the result will be a circular dependency, which should always be avoided.

As the status page does not have a public API anymore, the status-page frontend for a specific mesh protocol will always be provided by the same package the backend comes in. As such, there is no common functionality in different backend packages that would justify a PROVIDES.

File: package/gluon-status-page/files/lib/gluon/status-page/www/cgi-bin/dyn/neighbours-nodeinfo (hunk 3 above)
Comment (christf) [reply package/gluon-status-page/files/lib/gluon/status-page/www/cgi-bin/dyn/neighbours-nodeinfo]: agreed - how about we create gluon-get-mesh-ifs  that contains everything up to the Igrep. This would be usable in a more universal way than when incorporating the grep for a specific interface....

File: package/gluon-status-page-mesh-batman-adv/Makefile (hunk 2 above)
Comment (christf) [reply package/gluon-status-page-mesh-batman-adv/Makefile]: done

File: package/gluon-status-page/files/lib/gluon/status-page/www/cgi-bin/dyn/neighbours-nodeinfo (hunk 3 above)
Comment (christf) [reply package/gluon-status-page/files/lib/gluon/status-page/www/cgi-bin/dyn/neighbours-nodeinfo]: done

File: package/gluon-status-page/src/stations.c (hunk 5 above)
Comment (christf) [reply package/gluon-status-page/src/stations.c]: in general I agree but since we are splitting backend and frontend here, I left it in the commit. I will remove it now, but I really do like those stats.

File: package/gluon-status-page-mesh-batman-adv/Makefile (hunk 6 above)
Comment (christf) [reply package/gluon-status-page-mesh-batman-adv/Makefile]: done.

File: package/gluon-core/files/usr/bin/get-mesh-ifs [hunk 8]
```diff
@@ -0,0 +1,2 @@
+#!/bin/sh
+ubus call network.interface dump | jsonfilter -e "@.interface[@.proto='gluon_mesh' && @.up=true].device"
```
Comment (NeoRaider) [inline package/gluon-core/files/usr/bin/get-mesh-ifs]: Maybe name it "gluon-mesh-interfaces"? In any case, it should start with "gluon-".

File: package/gluon-l3roamd/files/etc/init.d/gluon-l3roamd [hunk 9]
```diff
@@ -10,7 +10,7 @@ start_service () {
 	procd_set_param stdout 1
 	procd_set_param stderr 1
 	procd_set_param respawn ${respawn_threshold:-3660} ${respawn_timeout:-5} ${respawn_retry:-0}
-	interfaces=$(for dev in $( ubus call network.interface dump | jsonfilter -e "@.interface[@.proto='gluon_mesh' && @.up=true].device");do echo " -m $dev"; done;
+	interfaces=$(for dev in $(/usr/bin/get-mesh-ifs);do echo " -m $dev"; done;
```
Comment (NeoRaider) [inline package/gluon-l3roamd/files/etc/init.d/gluon-l3roamd]: Remove the full path here, just use the name of the script.

**Comment by NeoRaider** [comment]
I've prepared the frontend part in 35ade80f1c827fd5d61528152a8a4e58ceccc7fa.

File: package/gluon-core/files/usr/bin/get-mesh-ifs (hunk 8 above)
Comment (christf) [reply package/gluon-core/files/usr/bin/get-mesh-ifs]: sure. I like verbs, so I would go with gluon-get-mesh-interfaces ok?

Comment (NeoRaider) [reply package/gluon-core/files/usr/bin/get-mesh-ifs]: Ok

File: package/gluon-l3roamd/files/etc/init.d/gluon-l3roamd (hunk 9 above)
Comment (christf) [reply package/gluon-l3roamd/files/etc/init.d/gluon-l3roamd]: removed the path here but kept it in the scripts that are called by the statuspage.

Comment (NeoRaider) [reply package/gluon-l3roamd/files/etc/init.d/gluon-l3roamd]: Why, do those script not have a proper PATH set?

**Comment by christf** [comment]
rename and rebase is done.

File: package/gluon-status-page/files/lib/gluon/status-page/www/cgi-bin/dyn/neighbours-nodeinfo [hunk 10]
```diff
@@ -6,6 +6,6 @@ badrequest() {
         exit 1
 }
 
-( ubus call network.interface dump | jsonfilter -e "@.interface[@.proto='gluon_mesh' && @.up=true].device"| grep -qxF "$QUERY_STRING" ) 2>/dev/null || badrequest
+( /usr/bin/gluon-get-mesh-interfaces | grep -qxF "$QUERY_STRING" ) 2>/dev/null || badrequest
```
Comment (NeoRaider) [inline package/gluon-status-page/files/lib/gluon/status-page/www/cgi-bin/dyn/neighbours-nodeinfo]: The other absolute paths in these files don't make much sense either, let's not introduce another one.

Comment (NeoRaider) [reply package/gluon-status-page/files/lib/gluon/status-page/www/cgi-bin/dyn/neighbours-nodeinfo]: Sorry about changing my mind again, but I think I would prefer the verb "list" to "get" in the script name.

Comment (christf) [reply package/gluon-status-page/files/lib/gluon/status-page/www/cgi-bin/dyn/neighbours-nodeinfo]: done. - for the paths I had to touch the files again anyways. I like list better than get too.

**Comment by NeoRaider** [comment]
Thanks, merged.
